expression/expression.go

```diff
@@ -173,6 +173,20 @@ func IsEQCondFromIn(expr Expression) bool {
 	return len(cols) > 0
 }
 
+// NotNull checks if an expression is constantly not null.
+func NotNull(l Expression) bool {
```

Comment:
This function would be more powerful. Like binary operators: col op col (const). For example 
```
mysql> create table ss (a int not null, b int);
mysql> create table tt (a int not null, b int);

mysql> explain select  ((t.a,t.b) in (select a,b from ss as s)) from tt as t;
+-------------------------+----------+-----------+---------------------------------------------------------------------------------------------------------------+
| id                      | count    | task      | operator info                                                                                                 |
+-------------------------+----------+-----------+---------------------------------------------------------------------------------------------------------------+
| Projection_6            | 10000.00 | root      | Column#9                                                                                                      |
| └─HashLeftJoin_7        | 10000.00 | root      | left outer semi join, inner:TableReader_11, equal:[eq(Column#1, Column#4)], other cond:eq(Column#2, Column#5) |
|   ├─TableReader_9       | 10000.00 | root      | data:TableScan_8                                                                                              |
|   │ └─TableScan_8       | 10000.00 | cop[tikv] | table:t, range:[-inf,+inf], keep order:false, stats:pseudo                                                    |
|   └─TableReader_11      | 10000.00 | root      | data:TableScan_10                                                                                             |
|     └─TableScan_10      | 10000.00 | cop[tikv] | table:s, range:[-inf,+inf], keep order:false, stats:pseudo                                                    |
+-------------------------+----------+-----------+---------------------------------------------------------------------------------------------------------------+
6 rows in set (0.00 sec)

mysql> explain select  ((t.a+1,t.b) in (select a,b from ss as s)) from tt as t;
+-------------------------+----------+-----------+--------------------------------------------------------------------------------------------------------------------------+
| id                      | count    | task      | operator info                                                                                                            |
+-------------------------+----------+-----------+--------------------------------------------------------------------------------------------------------------------------+
| Projection_6            | 10000.00 | root      | Column#9                                                                                                                 |
| └─HashLeftJoin_7        | 10000.00 | root      | CARTESIAN left outer semi join, inner:TableReader_11, other cond:eq(Column#2, Column#5), eq(plus(Column#1, 1), Column#4) |
|   ├─TableReader_9       | 10000.00 | root      | data:TableScan_8                                                                                                         |
|   │ └─TableScan_8       | 10000.00 | cop[tikv] | table:t, range:[-inf,+inf], keep order:false, stats:pseudo                                                               |
|   └─TableReader_11      | 10000.00 | root      | data:TableScan_10                                                                                                        |
|     └─TableScan_10      | 10000.00 | cop[tikv] | table:s, range:[-inf,+inf], keep order:false, stats:pseudo                                                               |
+-------------------------+----------+-----------+--------------------------------------------------------------------------------------------------------------------------+
6 rows in set (0.00 sec)
```
These second can avoid the CARTESIAN join.

Reply:
Yes,I should consider scalar function.